src/ffi_util.rs

```diff
@@ -198,8 +198,14 @@ impl<'a> CStrLike for &'a CString {
 /// Owned malloc-allocated memory slice.
 /// Do not derive `Clone` for this because it will cause double-free.
 pub struct CSlice {
-    pub data: *const c_char,
-    pub len: size_t,
+    data: *const c_char,
+    len: size_t,
+}
+
+impl CSlice {
+    pub(crate) fn new(data: *const c_char, len: size_t) -> Self {
```

Comment:
All good now, as far as library users are concerned, which is ultimately all that matters. Could still make this unsafe, so that even internally constructing CSlice's requires appropriate care.

Reply:
I've added some more comment on the constructor, and also marked it as unsafe, which gives more warning to internal uses.